staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go

```diff
@@ -133,3 +136,167 @@ func TestWatchErrResultNotBlockAfterCancel(t *testing.T) {
+			watchKey:         "/pods/",
+			watcherMaxLimit:  2,
+			expectEventCount: 2,
+			expectGetCount:   1,
+		},
+		{
+			name:             "a new item added to etcd before returning a second page is not returned: sync with two page",
+			watchKey:         "/pods/",
+			watcherMaxLimit:  1,
+			expectEventCount: 2,
+			expectGetCount:   2,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			defaultWatcherMaxLimit = testCase.watcherMaxLimit
+
+			kvWrapper := newEtcdClientKVWrapper(store.client.KV, store)
+			store.client.KV = kvWrapper
+
+			w := store.watcher.createWatchChan(
+				origCtx,
+				testCase.watchKey,
+				0,
+				true,
+				false,
+				newTestTransformer(),
+				storage.Everything)
+
+			err = w.sync()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// close incomingEventChan so we can read incomingEventChan non-blocking
+			close(w.incomingEventChan)
+
+			watched := 0
+			for {
+				event, ok := <-w.incomingEventChan
+				if !ok {
+					break
+				}
+
+				storagetesting.ExpectContains(t, "incorrect list pods", initList, event.key)
+				watched++
+			}
+
+			if watched != testCase.expectEventCount {
+				t.Errorf("Unexpected number of events: %v, expected: %v", watched, testCase.expectEventCount)
+			}
+
+			if kvWrapper.getCallCounter != testCase.expectGetCount {
+				t.Errorf("Unexpected called times of client.KV.Get() : %v, expected: %v", kvWrapper.getCallCounter, testCase.expectGetCount)
+			}
+		})
+	}
+}
+
+type etcdClientKVWrapper struct {
+	clientv3.KV
+	store *store
+	// keeps track of the number of times Get method is called
+	getCallCounter int
+	// getPostCall is called after the etcd KV's get function is executed.
+	getPostCall func(store *store) error
+}
+
+func newEtcdClientKVWrapper(kv clientv3.KV, store *store) *etcdClientKVWrapper {
+	return &etcdClientKVWrapper{
+		KV:             kv,
+		store:          store,
+		getCallCounter: 0,
+		getPostCall:    getPostCall,
+	}
+}
+
+func (ecw *etcdClientKVWrapper) Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error) {
+	resp, err := ecw.KV.Get(ctx, key, opts...)
+	ecw.getCallCounter++
+	if err != nil {
+		return nil, err
+	}
+
+	// execute getPostCall after the first call.
+	if ecw.getCallCounter == 1 {
+		if err := ecw.getPostCall(ecw.store); err != nil {
+			return nil, err
+		}
+	}
+
+	return resp, nil
+}
+
+func getPostCall(store *store) error {
+	barThird := &example.Pod{ObjectMeta: metav1.ObjectMeta{Namespace: "third", Name: "bar"}}
+	podKey := fmt.Sprintf("/pods/%s/%s", barThird.Namespace, barThird.Name)
+	storedObj := &example.Pod{}
+
+	err := store.Create(context.Background(), podKey, barThird, storedObj, 0)
+	if err != nil {
+		return fmt.Errorf("failed to create object: %w", err)
+	}
+	return nil
+}
+
+func initStoreData(ctx context.Context, store storage.Interface) ([]interface{}, error) {
```

Comment:
Let's not return interface - let's explicitly return []string if we just care about the keys...

Reply:
I  want to use `func ExpectContains(t *testing.T, msg string, expectedList []interface{}, got interface{})` to verfify the event, so return `[]interface{}`